src/Microsoft.ML.Api/DataViewConstructionUtils.cs

```diff
@@ -129,112 +129,192 @@ private Delegate CreateGetter(int index)
+                    var peek = DataView._peeks[index] as Peek<TRow, TSrc[]>;
                     Ch.AssertValue(peek);
-
-                    string[] buf = null;
-
-                    return (ValueGetter<VBuffer<DvText>>)((ref VBuffer<DvText> dst) =>
+                    TSrc[] buf = default;
+                    return (ValueGetter<VBuffer<TDst>>)((ref VBuffer<TDst> dst) =>
                     {
                         peek(GetCurrentRowObject(), Position, ref buf);
                         var n = Utils.Size(buf);
-                        dst = new VBuffer<DvText>(n, Utils.Size(dst.Values) < n
-                            ? new DvText[n]
+                        dst = new VBuffer<TDst>(n, Utils.Size(dst.Values) < n
+                            ? new TDst[n]
                             : dst.Values, dst.Indices);
                         for (int i = 0; i < n; i++)
-                            dst.Values[i] = new DvText(buf[i]);
-                    });
-                }
-
-                private Delegate CreateStringToTextGetter(int index)
-                {
-                    var peek = DataView._peeks[index] as Peek<TRow, string>;
-                    Ch.AssertValue(peek);
-                    string buf = null;
-                    return (ValueGetter<DvText>)((ref DvText dst) =>
-                        {
-                            peek(GetCurrentRowObject(), Position, ref buf);
-                            dst = new DvText(buf);
-                        });
-                }
-
-                private Delegate CreateBooleanToDvBoolGetter(int index)
-                {
-                    var peek = DataView._peeks[index] as Peek<TRow, bool>;
-                    Ch.AssertValue(peek);
-                    bool buf = false;
-                    return (ValueGetter<DvBool>)((ref DvBool dst) =>
-                    {
-                        peek(GetCurrentRowObject(), Position, ref buf);
-                        dst =  (DvBool)buf;
+                            dst.Values[i] = convert(buf[i]);
                     });
                 }
 
-                private Delegate CreateNullableBooleanToDvBoolGetter(int index)
+                private Delegate CreateGetterDelegate<TSrc, TDst>(int index, Func<TSrc, TDst> convert)
                 {
-                    var peek = DataView._peeks[index] as Peek<TRow, bool?>;
+                    var peek = DataView._peeks[index] as Peek<TRow, TSrc>;
                     Ch.AssertValue(peek);
-                    bool? buf = null;
-                    return (ValueGetter<DvBool>)((ref DvBool dst) =>
+                    TSrc buf = default;
+                    return (ValueGetter<TDst>)((ref TDst dst) =>
                     {
                         peek(GetCurrentRowObject(), Position, ref buf);
-                        dst = buf.HasValue ? (DvBool)buf.Value : DvBool.NA;
+                        dst = convert(buf);
```

Comment:
>convert [](start = 30, length = 7)

If `convert` is an identity, we are still incurring a delegate method call here, which is not efficient.

Reply:
I don't have identity converts. even if I pass  is x=>x I convert from TSrc to TDst class and call implicit converter. Which I'm not sure how to express through generic classes.
I would like to get rid of them, I just don't know way how to do it.

---
In reply to: [203875078](https://github.com/dotnet/machinelearning/pull/555#discussion_r203875078) [](ancestors = 203875078)

Reply:
Oh I see now. Ok, why don't we call this `CreateConvertingGetterDelegate`, to make it clear that there's also `CreateDirectGetterDelegate`?

---
In reply to: [203876308](https://github.com/dotnet/machinelearning/pull/555#discussion_r203876308) [](ancestors = 203876308,203875078)